Consider making API more robust by returning enum from `len`

One thing I found when using this API is that it's easy to accidentally forget handling overflows. Yes, `try_into_bytes` returns a result, but you don't actually have to call that (the data has already been written into the wrapped slice, where you need it).

One idea that would make my use case more robust:
1. Rename `len` to `bytes_written` (optional, but I think keeping `len` while implementing the rest of this suggestions would cause confusion).
2. Return enum `BytesWritten` from `bytes_written`, with the variants `BytesWritten::Nominal(usize)` and `BytesWritten::Overflow(usize)` (not sure if those are the best names). That way, the user is always forced to check for overflow.

What do you think?

If this problem exists it needs to be resolve by other structures similar to `heapless::Vec<_, _>` where you don't have directly access to memory. One of the aims of static-bytes is to resolve boilerplate. returning enum from `bytes_written()` will make this as complicated as at the beginning. But `bytes_written()` is a much much better name than `len()` and I will rename it anyway.

I have tried to test your problem:
```    #[test]
    fn issue_1() {
        let mut static_data = [0u8; 32];
        let mut safe_slice = SafeBytesSlice::from(&mut static_data[..]);

        fill_with_random(&mut safe_slice, 32);
        assert_eq!(safe_slice.is_exceed(), false);

        let _output = &static_data[..safe_slice.len()];
    }
```
And I have got compile-time error:
```
error[E0502]: cannot borrow `static_data` as immutable because it is also borrowed as mutable
   --> src/bytes_slice.rs:225:24
    |
220 |         let mut safe_slice = SafeBytesSlice::from(&mut static_data[..]);
    |                                                        ----------- mutable borrow occurs here
...
225 |         let _output = &static_data[..safe_slice.len()];
    |                        ^^^^^^^^^^^   ---------- mutable borrow later used here
    |                        |
    |                        immutable borrow occurs here

error: aborting due to previous error
```

Can you provide a minimal example when rust allows your incorrect code to compile?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,12 +18,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Security:
 - ...
 
+## [0.2.0] - 2020-11-01
+### Changed
+- Use 0.6.X API version of bytes crate
 
 ## [0.1.0] - 2020-08-29
 ### Added
 - Init version of `SafeBytesSlice`
 
-[Unreleased]: https://github.com/xoac/static-bytes/compare/v0.1.0...HEAD
+[Unreleased]: https://github.com/xoac/static-bytes/compare/v0.2.0...HEAD
+[0.2.0]: https://github.com/xoac/static-bytes/releases/tag/v0.2.0
 [0.1.0]: https://github.com/xoac/static-bytes/releases/tag/v0.1.0
 
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,7 +1,7 @@
 [package]
 name = "static-bytes"
 description = "Bytes for embedded devices"
-version = "0.1.0"
+version = "0.2.0"
 authors = ["Sylwester Rąpała <sylwesterrapala@outlook.com>"]
 edition = "2018"
 license = "MIT OR Apache-2.0"
@@ -50,7 +50,7 @@ categories = ["no-std", "embedded", "data-structures"]
  maintenance = { status = "experimental" }
 
 [dependencies]
-bytes = { version = "0.5", default-features=false }
+bytes = { version = "^0.6", default-features=false }
 
 
 # This Cargo.toml was generated from template: https://github.com/xoac/crates-io-lib-template

--- a/README.md
+++ b/README.md
@@ -20,7 +20,13 @@ There are two problems with code above:
 - it will panic if encode want to use more than 16 bytes!
 - it is boilerplate
 
-You can resolve both with `SafeBytesSlice`. For example usage see docs.
+You can resolve both with `SafeBytesSlice`. For example usage see
+[docs](https://docs.rs/static-bytes/0.2/static_bytes/struct.SafeBytesSlice.html).
+
+### Compatibility with bytes
+- v0.1.x is compatible with bytes >=0.5.0, <0.6.0
+- v0.2.x is compatible with bytes >=0.6.0, <0.7.0
+
 
 ## License
 

--- a/src/bytes_slice.rs
+++ b/src/bytes_slice.rs
@@ -174,7 +174,7 @@ unsafe impl<'a> BufMut for SafeBytesSlice<'a> {
         let bytes = &mut self.slice[self.bytes_written..];
         let len = bytes.len();
         let ptr = bytes.as_mut_ptr() as *mut _;
-        unsafe { return UninitSlice::from_raw_parts_mut(ptr, len) }
+        unsafe { UninitSlice::from_raw_parts_mut(ptr, len) }
     }
 
     fn put_slice(&mut self, src: &[u8]) {

--- a/src/bytes_slice.rs
+++ b/src/bytes_slice.rs
@@ -1,4 +1,5 @@
 use super::error::CapacityExceeded;
+use bytes::buf::UninitSlice;
 use bytes::BufMut;
 use core::mem;
 use core::mem::MaybeUninit;
@@ -153,7 +154,7 @@ impl<'a> SafeBytesSlice<'a> {
 }
 
 // Implement required methods
-impl<'a> BufMut for SafeBytesSlice<'a> {
+unsafe impl<'a> BufMut for SafeBytesSlice<'a> {
     fn remaining_mut(&self) -> usize {
         debug_assert!(self.bytes_written <= self.slice.len());
         self.slice.len() - self.bytes_written
@@ -169,22 +170,28 @@ impl<'a> BufMut for SafeBytesSlice<'a> {
         }
     }
 
-    fn bytes_mut(&mut self) -> &mut [mem::MaybeUninit<u8>] {
-        &mut self.slice[self.bytes_written..]
+    fn bytes_mut(&mut self) -> &mut UninitSlice {
+        let bytes = &mut self.slice[self.bytes_written..];
+        let len = bytes.len();
+        let ptr = bytes.as_mut_ptr() as *mut _;
+        unsafe { return UninitSlice::from_raw_parts_mut(ptr, len) }
     }
 
     fn put_slice(&mut self, src: &[u8]) {
         use core::ptr;
         // check if we have enough data to put slice. If no set flag instead of panic!
-        if self.remaining_mut() < src.len() {
+        let src_len = src.len();
+        if self.remaining_mut() < src_len {
+            self.bytes_written = self.slice.len(); // make `remaining_mut()` return 0
             self.cap_exceeded = true;
             return;
         }
 
+        // enough inner capacity to execute safe copy
         unsafe {
             let dst = self.bytes_mut();
-            ptr::copy_nonoverlapping(src[..].as_ptr(), dst.as_mut_ptr() as *mut u8, src.len());
-            self.advance_mut(src.len());
+            ptr::copy_nonoverlapping(src[..].as_ptr(), dst.as_mut_ptr() as *mut u8, src_len);
+            self.advance_mut(src_len);
         }
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -13,7 +13,13 @@
 //! - it will panic if encode want to use more than 16 bytes!
 //! - it is boilerplate
 //!
-//! You can resolve both with `SafeBytesSlice`. For example usage see docs.
+//! You can resolve both with `SafeBytesSlice`. For example usage see
+//! [docs](https://docs.rs/static-bytes/0.2/static_bytes/struct.SafeBytesSlice.html).
+//!
+//! ## Compatibility with bytes
+//! - v0.1.x is compatible with bytes >=0.5.0, <0.6.0
+//! - v0.2.x is compatible with bytes >=0.6.0, <0.7.0
+//!
 
 #![deny(missing_docs)]
 #![no_std]

